src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,607 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <string.h>
+
+#include "include/buffer.h"
+#include "include/byteorder.h"
+#include "crimson/os/seastore/transaction_manager.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "seastar/core/thread.hh"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::omap_manager {
+std::ostream &operator<<(std::ostream &out, const omap_inner_key_t &rhs)
+{
+  return out << "omap_inner_key (" << rhs.key_off<< " - " << rhs.key_len
+             << " - " << rhs.laddr << ")";
+}
+
+std::ostream &operator<<(std::ostream &out, const omap_leaf_key_t &rhs)
+{
+  return out << "omap_leaf_key_t (" << rhs.key_off<< " - " << rhs.key_len
+             << " "<< rhs.val_off<<" - " << rhs.val_len << ")";
+}
+
+std::ostream &OMapInnerNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << get_meta().depth;
+}
+
+OMapInnerNode::get_value_ret
+OMapInnerNode::get_value(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {} key = {}", "OMapInnerNode", __func__, key);
+  auto child_pt = get_containing_child(key);
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+    [oc, &key] (auto extent) {
+    return extent->get_value(oc, key);
+  });
+}
+
+OMapInnerNode::insert_ret
+OMapInnerNode::insert(omap_context_t oc, const std::string &key, const std::string &value)
+{
+  logger().debug("{}: {}  {}->{}", "OMapInnerNode",  __func__, key, value);
+  auto child_pt = get_containing_child(key);
+  assert(child_pt != iter_end());
+  auto laddr = child_pt->get_node_key().laddr;
+  return omap_load_extent(oc, this, laddr, get_meta().depth - 1).safe_then(
+    [this, oc, child_pt, &key, &value] (auto extent) {
+    auto size = extent->get_node_meta().depth > 1? key.size() + 1:
+                                           key.size() + value.size() + 2;
```

Comment:
This is the second place with this logic duplicated.  It could still probably be encoded in the extent_will_overflow() overload.